### PR TITLE
#3544 remove template references from IContent

### PR DESCRIPTION
**Pull request by StefanoChiodino**
### Prerequisites

- [X] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes: https://github.com/umbraco/Umbraco-CMS/issues/3544

### Description

@AnthonyCogworks  and @warrenbuckley helped me out with this during the UK fest hackathon 2018.

I've tried creating the an homepage, adding it to the content tree, adding another template, change template, etc. All seems to be working fine!

I've had to make it a nullable int, because you should be able to not set the template. In some instances I've noticed 0 was assumed to be "not set", however using that caused problems in the DB, because it assumed that a template with id 0 exists. A nullable makes more sense to me anyway.

I've given `PublishedTemplate` the same treatment, as I assumed the same problem applied there.

Still working on a few unit tests...

### CODE REVIEW

File: src/Umbraco.Core/Persistence/Repositories/Implement/DocumentRepository.cs [hunk 1]
```diff
@@ -405,7 +405,7 @@ protected override void PersistNewItem(IContent entity)
             else if (content.PublishedState == PublishedState.Unpublishing)
             {
                 content.Published = false;
-                content.PublishTemplate = null;
+                content.PublishTemplateId = 0;
```
Comment (stevetemple) [inline src/Umbraco.Core/Persistence/Repositories/Implement/DocumentRepository.cs]: Given that it was a null and we're using a Nullable<int> shouldn't we keep it as null

File: src/Umbraco.Web/Models/Mapping/DefaultTemplateResolver.cs [hunk 2]
```diff
@@ -8,17 +10,24 @@ internal class DefaultTemplateResolver : IValueResolver<IContent, ContentItemDis
...
-            //set default template if template isn't set
-            if (string.IsNullOrEmpty(alias))
-                alias = source.ContentType.DefaultTemplate == null
-                    ? string.Empty
-                    : source.ContentType.DefaultTemplate.Alias;
+                // ... otherwise return the content type default template alias.
+                return source.ContentType.DefaultTemplate?.Alias;
+            }
+            
+            var fileService = DependencyResolver.Current.GetService<IFileService>();
```
Comment (bergmania) [inline src/Umbraco.Web/Models/Mapping/DefaultTemplateResolver.cs]: I and some others think this is a anti pattern (http://blog.ploeh.dk/2010/02/03/ServiceLocatorisanAnti-Pattern/)
You should be able to inject IFileService in the constructor here and in `ContentMapperProfile`. Then you just use LightInject to handle it.

File: src/Umbraco.Core/Persistence/Repositories/Implement/DocumentRepository.cs [hunk 3]
```diff
@@ -266,8 +266,8 @@ protected override void PersistNewItem(IContent entity)
             var publishing = content.PublishedState == PublishedState.Publishing;
 
             // ensure that the default template is assigned
-            if (entity.Template == null)
-                entity.Template = entity.ContentType.DefaultTemplate;
+            if (entity.TemplateId == 0)
```
Comment (bergmania) [inline src/Umbraco.Core/Persistence/Repositories/Implement/DocumentRepository.cs]: You should be able to handle it like the nullable it is:

`if (entity.TemplateId == null)
                entity.Template = entity.ContentType.DefaultTemplate;	                entity.TemplateId = entity.ContentType.DefaultTemplate?.Id;`

File: src/Umbraco.Core/Persistence/Repositories/Implement/DocumentRepository.cs [hunk 4]
```diff
@@ -605,7 +605,7 @@ protected override void PersistUpdatedItem(IContent entity)
             else if (content.PublishedState == PublishedState.Unpublishing)
             {
                 content.Published = false;
-                content.PublishTemplate = null;
+                content.PublishTemplateId = 0;
```
Comment (bergmania) [inline src/Umbraco.Core/Persistence/Repositories/Implement/DocumentRepository.cs]: again, `null` should be used

File: src/Umbraco.Tests/Persistence/Repositories/DocumentRepositoryTest.cs [hunk 5]
```diff
@@ -557,12 +558,12 @@ public void UpdateContentWithNullTemplate()
                 var repository = CreateRepository((IScopeAccessor)provider, out _);
 
                 var content = repository.Get(NodeDto.NodeIdSeed + 2);
-                content.Template = null;
+                content.TemplateId = 0;
```
Comment (bergmania) [inline src/Umbraco.Tests/Persistence/Repositories/DocumentRepositoryTest.cs]: again, null should be used, and the we also need to update the assert to use `Assert.IsNull`

File: src/Umbraco.Web/Models/Mapping/DefaultTemplateResolver.cs (hunk 2 above)
Comment (StefanoChiodino) [reply src/Umbraco.Web/Models/Mapping/DefaultTemplateResolver.cs]: @bergmania I tried ctor and parameter injection and they didn't work, unfortunately!

**Comment by StefanoChiodino** [comment]
After a bit more tweaking all tests pass 🎊

**Pull request merged**